packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,279 @@
+      this.stopPoller()
+      return
+    } else {
+      this.retryCount++
+      this.lastError = new Error(`[${response.status}] ${response.statusText}`)
+      timeUntilRefreshInMs = backoff({
+        attempt: this.retryCount,
+        minTimeout: 25,
+        maxTimeout: 1000,
+      })
+    }
+
+    if (this.retryCount % this.maxRetries === 0) {
+      this.tokenEmitter.emit('access_token', { error: this.lastError })
+      // TODO: figure out timing and whether to reset retries?
+    }
+    this.pollerTimer = setTimeout(
+      this.pollerLoop.bind(this),
+      timeUntilRefreshInMs
+    )?.unref()
+  }
+
+  stopPoller() {
+    clearTimeout(this.pollerTimer)
+  }
+
+  /**
+   * Solely responsible for building the HTTP request and calling the token service.
+   */
+  private requestAccessToken(): Promise<HTTPResponse> {
+    const jti = uuid()
+    const currentUTCInSeconds = Math.round(Date.now() / 1000)
+    const jwtBody = {
+      iss: this.clientId,
+      sub: this.clientId,
+      aud: this.authServer,
+      iat: currentUTCInSeconds,
+      exp: currentUTCInSeconds + 60,
+      jti,
+    }
+
+    const signingOptions: SignOptions = {
+      algorithm: this.alg,
+      keyid: this.keyId,
+    }
+
+    const signedJwt = sign(jwtBody, this.clientKey, signingOptions)
+
+    const requestBody = `grant_type=${this.grantType}&client_assertion_type=${this.clientAssertionType}&client_assertion=${signedJwt}&scope=${this.scope}`
+    const accessTokenEndpoint = `${this.authServer}/token`
+
+    const requestOptions: HTTPClientRequest = {
+      method: 'POST',
+      url: accessTokenEndpoint,
+      body: requestBody,
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      httpRequestTimeout: 10000,
+    }
+    return this.httpClient.makeRequest(requestOptions)
+  }
+
+  async getAccessToken(): Promise<AccessToken> {
+    // Use the cached token if it is still valid, otherwise wait for a new token.
+    if (this.isValidToken(this.accessToken)) {
+      return this.accessToken
+    }
+
+    // stop poller first in order to make sure that it's not sleeping if we need a token immediately
+    // Otherwise it could be hours before the expiration time passes normally
+    this.stopPoller()
+
+    // startPoller needs to be called somewhere, either lazily when a token is first requested, or at instantiation.
+    // Doing it lazily currently
+    this.pollerLoop().catch(() => {})
+
+    return new Promise((resolve, reject) => {
+      this.tokenEmitter.once('access_token', (event) => {
+        if ('token' in event) {
+          resolve(event.token)
+        } else {
+          reject(event.error)
+        }
+      })
+    })
+  }
+
+  clearToken() {
+    this.accessToken = undefined
+  }
+
+  isValidToken(token?: AccessToken): token is AccessToken {
+    // TODO: Check if it has already expired?
+    // otherwise this check is pretty much useless
+    return typeof token !== 'undefined' && token !== null
```

Comment:
Would we want to rename this to "hasAccessToken"? Seems a little confusing to me that this is called "isValidToken" but we already have a isAccessToken helper.

or, can we just completely remove this and just do "if (this.accessToken)" where appropriate?

Reply:
Yeah I think that's reasonable. Comment suggests as much :smile:

Reply:
Added expiration check instead.